### ISSUE TITLE
Post transform query cache

### ISSUE THREAD

**Issue by gtk-grafana** [grafana/grafana#69807]
In order to prepare for the migration of [transformV2 to the backend](https://github.com/grafana/grafana/issues/69018), incremental query needs to cache post-transform values.

This should also reduce the amount of computation we run on each incremental query.

**Comment by gtk-grafana** [comment]
So @leeoniya and I did some investigating around this yesterday. We found a funky looking heatmap that was generated after the ops-cortex-dev prometheus instance crashed/rebooted yesterday:
<img width="1671" alt="image" src="https://github.com/grafana/grafana/assets/109082771/1d08d2b1-a5a4-47a2-9a3e-46a46af96fda">
Every quantile value should be less then the first +Inf value, and also histogram data frames are expected to be square (they all have the same number of values)
![image](https://github.com/grafana/grafana/assets/109082771/c8c80175-11bc-43f6-9ba0-dd9e4a74bbea)

While this non-square dataframe does cause post-transform incremental query to mangle the values, I think this is acceptable because a non-square dataframe isn't a valid heatmap.

TL;DR: There is a bug if heatmaps can have series of different lengths, but AFAIK this is a fundamental identity of heatmap dataframes, so it's not worth trying to null-fill or join the series values to get incremental querying to work.

**Comment by gtk-grafana** [comment]
The result of this investigation is that incremental querying should continue to work when the transform is moved to the backend. Still be good to test before merging in though
